Add a test checking that all the C headers are also compatible with C++

kit-ty-kate opened the pull request:
C++ projects have been broken quite often due to the C headers not being checked with a C++ compiler.
This PR fixes all problems in the headers as well as add a test checking all headers to be sure they are compatible with C++ in the future.

As you can see in the last commit, this also fixed an issue that made C++ compiler break when using `CAML_INTERNALS`. This case can be seen in the wild here: https://github.com/ocaml/opam-repository/pull/22140#issuecomment-1254927737

*Side note: I've been trying to make this branch work a year and a half ago and never could make it succeed for some reason, but somehow i tried it today again and it works fine with trunk now. I've been trying to be careful when rebasing so hopefully I haven't ported back some old behaviour to the driver somehow.*

kit-ty-kate commented:
I would also suggest to backport it to `5.0` as it is also broken there.
~If it is to be backported no changelog entry should be necessary I believe.~ EDIT: Actually scratch that part, the driver now accept .cpp files so that should be there.

dra27 commented:
[precheck#768](https://ci.inria.fr/ocaml/job/precheck/768/)

kit-ty-kate commented on driver/compenv.ml:
```diff
@@ -712,8 +725,15 @@ let process_deferred_actions env =
     match final_output_name with
     | None -> ()
-    | Some _output_name ->
+    | Some output_name ->
         if !compile_only then begin
+          if List.filter (function
+              | ProcessCFile name ->
+                c_object_of_filename ~ext:".c" name <> output_name
+              | ProcessCppFile name ->
+                c_object_of_filename ~ext:".cpp" name <> output_name
+              | _ -> false) !deferred_actions <> [] then
+            fatal "Options -c and -o are incompatible when compiling C files";
           if List.length (List.filter (function
               | ProcessCFile _
```
oops that's probably a rebasing artifact
```suggestion
    | Some _output_name ->
        if !compile_only then begin
          if List.length (List.filter (function
              | ProcessCFile _
              | ProcessCppFile _
```

Octachron commented:
To resume an offline discussion, backporting the header fixes to 5.0 is fine with me.
However, I think it is better to keep the tests and the cpp file support on trunk (since 5.0 will still benefit for further fixes if the test on trunk detect a problem).

nojb commented on driver/compenv.ml:
```diff
@@ -693,6 +704,8 @@ let action_of_file name =
     ProcessInterface name
   else if Filename.check_suffix name ".c" then
     ProcessCFile name
+  else if Filename.check_suffix name ".cpp" then
```
There are other common extensions for C++: `.cxx`, `.cc`, etc, do we want to support them here?
```suggestion
  else if Filename.check_suffix name ".cpp" then
```

nojb commented on driver/compenv.ml:
```diff
@@ -627,12 +627,13 @@ type deferred_action =
-let c_object_of_filename name =
-  Filename.chop_suffix (Filename.basename name) ".c" ^ Config.ext_obj
+let c_object_of_filename ~ext name =
+  Filename.chop_suffix (Filename.basename name) ext ^ Config.ext_obj
```
Could use `Filename.remove_extension` to avoid having to pass `~ext`.

xavierleroy commented:
Globally: I like the idea of the test but I'm a bit nervous with adding C++ compilation support to ocamlc and ocamlopt. It's always possible to do `gcc -I$(ocamlc -where) -c foo.cpp` to compile a C++ file that refers to OCaml's headers.

shindere commented:
Xavier Leroy (2022/09/23 03:06 -0700):
> Globally: I like the idea of the test but I'm a bit nervous with
> adding C++ compilation support to ocamlc and ocamlopt. It's always
> possible to do `gcc -I$(ocamlc -where) -c foo.cpp` to compile a C++
> file that refers to OCaml's headers.

But, doing so, you loose the ability to call the C compiler with the
right options. Isn't that a problem / risk?

xavierleroy commented:
> But, doing so, you loose the ability to call the C compiler with the right options. Isn't that a problem / risk?

We're not calling a C compiler in this situation, we're calling a C++ compiler that may be the same command as the C compiler.  But the right options for a C compilation may not be the right option for a C++ compilation.  Just passing the same options as for a C compilation is a hail Mary.  Eh, even choosing between `gcc` and `g++` (or `clang` and `clang++`) to compile C++ sources is not obvious.  I don't want the OCaml core system to get into this C++ business.

shindere commented:
Okay, sorry. I assumed that C++ compilers merely extended the set of
command-line options of their C counterpart, but if that assumption is
wrong, it's indeed not a good idea.

kit-ty-kate committed:
Test every public C headers using a C++ compiler

3 tests are defined:
- raw include
- include inside an extern "C"
- raw include after #define CAML_INTERNALS

xavierleroy commented on testsuite/tests/cpp-api/cpp_stubs3.c:
```diff
@@ -0,0 +1,17 @@
+#define CAML_INTERNALS
```
The CAML_INTERNALS sections of the OCaml 5 header files contain too much C code (as inline functions) to be compilable by a C++ compiler.  So, this 3rd test will never work, I'm afraid.  Let's just test without CAML_INTERNALS.

xavierleroy commented on testsuite/tests/cpp-api/cpp.ml:
```diff
@@ -0,0 +1,15 @@
+(* TEST
+   modules = "cpp_stubs1.c cpp_stubs2.c cpp_stubs3.c";
+   readonly_files = "all-includes.h";
+   flags = "-ccopt -x -ccopt c++ -ccopt -std=c++11";
```
These flags work with GCC and Clang, but probably not with MSVC.  Can the test be turned off for MSVC?

dra27 replied:
Yes - leftover from #13476, we have a `not-msvc` action!

xavierleroy reviewed (commented):
Thank you for resurrecting this PR, it's a useful test to have.
